### ISSUE TITLE
Pyright erroneously expecting fewer positional arguments than there are with dataclasses.

### ISSUE THREAD

**Issue by DemonicSavage** [microsoft/pyright#3721]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Pyright gives a reportGeneralTypeIssues issue when an object is created, that says "Expected n positional arguments" when n is a lower number than the actual number of positional arguments the class has.

**To Reproduce**
1. Create a dataclass
2. Create a new object using the correct amount of arguments.

**Expected behavior**
Pyright shouldn't give any issues.

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/1365501/179839211-1fb4dc17-6f0e-4c12-8e45-79bdc4e2d8bd.png)
![image](https://user-images.githubusercontent.com/1365501/179838937-5b26c1cf-85ad-4ea9-bf87-5a00a5d887a6.png)
![image](https://user-images.githubusercontent.com/1365501/179838972-1b2709f6-6349-4515-8bae-844281e70f00.png)
(import is there to avoid a circular import)

**VS Code extension or command-line**
VS Code, Pylance v2022.7.0

**Additional context**

**Comment by erictraut** [comment]
Could you please provide a self-contained minimal example in text form? Debugging issues from partial screen shots doesn't work well.

**Comment by DemonicSavage** [comment]
I cannot reproduce it in a minimal example, but sure:

```python
from dataclasses import dataclass
@dataclass
class Still:
    key: int
    url: str

x = Still(5, 'asa')
```

**Comment by DemonicSavage** [comment]
https://github.com/DemonicSavage/sifas_card_downloader/blob/main/src/html_parser.py#L172=

The error happens here, if you want code - I used `# type: ignore` for now.

**Comment by erictraut** [comment]
Here's a minimal example that repros the problem:

```python
from dataclasses import dataclass

@dataclass
class Still:
    key: int
    url: str

    def foo(self):
        self.url: str = ""

Still(1, "hi")
```

The issue is triggered by the presence of the redundant type annotation for `url`. Dataclasses are special in that you declare variables in the class body as though they are class variables, but these are turned into instance variables by the dataclass decorator. In your case, you are annotating `url` in two places — one in the class body (as though it's a class variable) and a second time in a method (as though it's an instance variable). This is what confused pyright.

You can work around the issue by eliminating the redundant annotation.

I'll investigate ways to make pyright more resilient in this case.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.262, which I just published. It will also be included in a future version of pylance.
